File: mobile-app/modules/sdk/ios/agent/AgentModule.swift
import ExpoModulesCore

/// Expo module for voice agent control (mirrors AgentModule.kt)
/// Thin layer that forwards calls to AgentManager and emits events to JS.
public class AgentModule: Module {

    private var notificationObserver: NSObjectProtocol?
    private var glassNotificationObserver: NSObjectProtocol?
    private var wakeupInactivityTask: Task<Void, Never>?
    private var isWakeupSessionActive = false
    private let wakeupTimeoutSeconds: Double = 10

    public func definition() -> ModuleDefinition {
        Name("AgentModule")

        Events(
            "AGENT_STATE_CHANGED",
            "AGENT_PARTICIPANT_EVENT",
            "AGENT_AUDIO_TRACK_EVENT",
            "AGENT_TEXT_EVENT",
            "AGENT_DATA_EVENT",
            "AGENT_ERROR",
            "AGENT_TOOL_CALL",
            "AGENT_TOOL_RESULT",
            "AGENT_ACTIVE_SPEAKERS",
            "AGENT_TOOL_EXECUTED",
            "AGENT_RPC_METHOD_CALLED"
        )

        OnCreate {
            SdkConfig.shared.initialize()
            ApiClient.shared.initialize(baseUrl: SdkConfig.shared.apiUrl)
            AgentManager.shared.initialize()
            ToolManager.shared.initialize()

            self.notificationObserver = NotificationCenter.default.addObserver(
                forName: .agentEvent,
                object: nil,
                queue: .main
            ) { [weak self] notification in
                guard let self, let msg = notification.userInfo?["msg"] as? AgentEventMsg else { return }
                // Inactivity auto-off handled by backend (exit_voice_mode) — app-side activity tracking disabled.
                // if self.isWakeupSessionActive {
                //     if msg.cmd == AgentEventConstants.msgActiveSpeakers,
                //        let s = msg.s1, !s.isEmpty, s != "[]" {
                //         self.resetWakeupTimer()
                //     } else if msg.cmd == AgentEventConstants.msgTextEvent {
                //         self.resetWakeupTimer()
                //     }
                // }
                self.emitAgentEvent(msg)
            }

            self.glassNotificationObserver = NotificationCenter.default.addObserver(
                forName: .glassEvent,
                object: nil,
                queue: .main
            ) { [weak self] notification in
                guard let msg = notification.userInfo?["msg"] as? GlassEventMsg else { return }
                switch msg.cmd {
                case GlassEventConstants.msgVoiceWakeup: self?.handleVoiceWakeup()
                case GlassEventConstants.msgVoiceStop:   self?.handleVoiceWakeupEnd()
                default: break
                }
            }
        }

        OnDestroy {
            if let observer = self.notificationObserver {
                NotificationCenter.default.removeObserver(observer)
                self.notificationObserver = nil
            }
            if let observer = self.glassNotificationObserver {
                NotificationCenter.default.removeObserver(observer)
                self.glassNotificationObserver = nil
            }
            self.wakeupInactivityTask?.cancel()
            self.isWakeupSessionActive = false
            AgentManager.shared.destroy()
        }

        // iOS has no OnActivityEntersForeground/Background.
        // The CameraTool accesses the root UIViewController directly via UIWindowScene.

        AsyncFunction("start") { (audio: Bool, promise: Promise) in
            Task {
                AgentManager.shared.start(audio: audio)
                promise.resolve(nil)
            }
        }

        AsyncFunction("stop") { (promise: Promise) in
            Task {
                AgentManager.shared.stop()
                promise.resolve(nil)
            }
        }

        AsyncFunction("sendText") { (text: String, promise: Promise) in
            Task {
                AgentManager.shared.sendText(text)
                promise.resolve(nil)
            }
        }

        AsyncFunction("sendFile") { (path: String, promise: Promise) in
            Task {
                do {
                    try await AgentManager.shared.sendFile(path)
                    promise.resolve(nil)
                } catch {
                    promise.reject(error)
                }
            }
        }

        AsyncFunction("sendData") { (data: Data, reliable: Bool, promise: Promise) in
            Task {
                AgentManager.shared.sendData(data, reliable: reliable)
                promise.resolve(nil)
            }
        }

        AsyncFunction("setMicEnabled") { (enabled: Bool, promise: Promise) in
            Task {
                AgentManager.shared.setMicEnabled(enabled)
                promise.resolve(nil)
            }
        }

        AsyncFunction("setSpeakerEnabled") { (enabled: Bool, promise: Promise) in
            Task {
                AgentManager.shared.setSpeakerEnabled(enabled)
                promise.resolve(nil)
            }
        }

        AsyncFunction("toggleAudio") { (promise: Promise) in
            Task {
                AgentManager.shared.toggleAudio()
                promise.resolve(nil)
            }
        }

        AsyncFunction("setAudioEnabled") { (enabled: Bool, promise: Promise) in
            Task {
                AgentManager.shared.setAudioEnabled(enabled)
                promise.resolve(nil)
            }
        }

        AsyncFunction("getConnectionState") { () -> String in
            return AgentManager.shared.getConnectionState().rawValue
        }

        AsyncFunction("getMicEnabled") { () -> Bool in
            return AgentManager.shared.getMicEnabled()
        }

        AsyncFunction("getSpeakerEnabled") { () -> Bool in
            return AgentManager.shared.getSpeakerEnabled()
        }

        AsyncFunction("getRoomId") { () -> String? in
            return AgentManager.shared.getRoomId()
        }

        AsyncFunction("startApp") { (appId: String, promise: Promise) in
            Task {
                AgentManager.shared.startApp(appId: appId)
                promise.resolve(nil)
            }
        }

        AsyncFunction("reloadTools") { (promise: Promise) in
            Task {
                AgentManager.shared.reloadTools()
                promise.resolve(nil)
            }
        }

        AsyncFunction("sendFiles") { (urls: [String], message: String, promise: Promise) in
            Task {
                do {
                    try await AgentManager.shared.sendFiles(urls, message: message)
                    promise.resolve(nil)
                } catch {
                    promise.reject(error)
                }
            }
        }

        AsyncFunction("performRpc") { (method: String, payload: String, promise: Promise) in
            Task {
                do {
                    let result = try await AgentManager.shared.performRpc(method: method, payload: payload)
                    promise.resolve(result)
                } catch {
                    promise.reject(error)
                }
            }
        }

        AsyncFunction("registerRpcMethod") { (method: String, promise: Promise) in
            Task {
                do {
                    try await AgentManager.shared.registerRpcMethod(method)
                    promise.resolve(nil)
                } catch {
                    promise.reject(error)
                }
            }
        }

        AsyncFunction("setCookie") { (cookie: String, promise: Promise) in
            AgentManager.shared.setCookie(cookie)
            promise.resolve(true)
        }

        AsyncFunction("setAgentName") { (name: String, promise: Promise) in
            AgentManager.shared.setAgentName(name.isEmpty ? nil : name)
            promise.resolve(nil)
        }

        AsyncFunction("startPhoneStream") { (fps: Int, bitrate: Int, promise: Promise) in
            Task {
                do {
                    try await AgentManager.shared.startPhoneStream(fps: fps, bitrate: bitrate)
                    promise.resolve(nil)
                } catch {
                    promise.reject("START_PHONE_STREAM_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("stopPhoneStream") { (promise: Promise) in
            AgentManager.shared.stopPhoneStream()
            promise.resolve(nil)
        }

        AsyncFunction("flipPhoneCamera") { (promise: Promise) in
            AgentManager.shared.flipPhoneCamera()
            promise.resolve(nil)
        }

        AsyncFunction("isPhoneStreaming") { () -> Bool in
            return AgentManager.shared.isPhoneStreaming()
        }

        AsyncFunction("startGlassStream") { (ssid: String, pwd: String, fps: Int, bitrate: Int, promise: Promise) in
            Task {
                do {
                    try await AgentManager.shared.startGlassStream(ssid: ssid, pwd: pwd, fps: fps, bitrate: bitrate)
                    promise.resolve(nil)
                } catch {
                    promise.reject("START_GLASS_STREAM_FAILED", error.localizedDescription)
                }
            }
        }

        AsyncFunction("stopGlassStream") { (promise: Promise) in
            AgentManager.shared.stopGlassStream()
            promise.resolve(nil)
        }

        AsyncFunction("isGlassStreaming") { () -> Bool in
            return AgentManager.shared.isGlassStreaming()
        }
    }

    // MARK: - Voice wakeup (mirrors AgentModule.kt handleVoiceWakeup / resetWakeupTimer)

    private func handleVoiceWakeup() {
        guard AgentManager.shared.getConnectionState() == .connected else { return }

        // Idempotent guard: ignore the wakeup whenever voice is already on — covers repeated
        // aiDialogue=true frames and the speaking_start echo, so we never re-send "hey" / re-enable.
        if AgentManager.shared.getMicEnabled() {
            return
        }

        AgentManager.shared.sendText("hey")

        if !AgentManager.shared.getMicEnabled() {
            AgentManager.shared.setAudioEnabled(true)
        }

        isWakeupSessionActive = true
        // Inactivity auto-off handled by backend (exit_voice_mode) — app-side timer disabled.
        // resetWakeupTimer()
    }

    /// Glasses' button turned voice mode OFF (device AI-dialogue falling edge) → turn off the
    /// phone-side voice and cancel the wakeup inactivity timer.
    private func handleVoiceWakeupEnd() {
        wakeupInactivityTask?.cancel()
        wakeupInactivityTask = nil
        isWakeupSessionActive = false
        if AgentManager.shared.getMicEnabled() {
            AgentManager.shared.setAudioEnabled(false)
        }
    }

    private func resetWakeupTimer() {
        wakeupInactivityTask?.cancel()
        wakeupInactivityTask = Task { [weak self] in
            guard let self else { return }
            try? await Task.sleep(for: .seconds(self.wakeupTimeoutSeconds))
            guard !Task.isCancelled else { return }
            AgentManager.shared.setAudioEnabled(false)
            self.isWakeupSessionActive = false
            self.wakeupInactivityTask = nil
        }
    }

    // MARK: - Event emission (mirrors AgentModule.kt emitEvent)

    private func emitAgentEvent(_ event: AgentEventMsg) {
        switch event.cmd {
        case AgentEventConstants.msgStateChanged:
            sendEvent("AGENT_STATE_CHANGED", ["state": event.s1 ?? ""])

        case AgentEventConstants.msgParticipantEvent:
            var data: [String: Any] = [
                "type": event.s1 ?? "",
                "participantId": event.s2 ?? "",
                "isSpeaking": event.n1 == 1,
                "isAgent": event.n2 == 1
            ]
            if let identity = event.s3 { data["identity"] = identity }
            if let name = event.s4     { data["name"] = name }
            sendEvent("AGENT_PARTICIPANT_EVENT", data)

        case AgentEventConstants.msgAudioTrackEvent:
            sendEvent("AGENT_AUDIO_TRACK_EVENT", [
                "type": event.s1 ?? "",
                "participantId": event.s2 ?? "",
                "trackId": event.s3 ?? "",
                "isLocal": event.n1 == 1,
                "isMuted": event.n2 == 1
            ])

        case AgentEventConstants.msgTextEvent:
            var data: [String: Any] = [
                "id": event.s1 ?? "",
                "text": event.s2 ?? "",
                "isFinal": event.n1 == 1,
                "isUser": event.n2 == 1,
                "isTranscription": event.n3 == 1
            ]
            if let seg = event.s3 { data["segmentId"] = seg }
            sendEvent("AGENT_TEXT_EVENT", data)

        case AgentEventConstants.msgDataEvent:
            var data: [String: Any] = [
                "participantId": event.s1 ?? "",
                "data": event.s2 ?? ""
            ]
            if let t = event.s3 { data["topic"] = t }
            sendEvent("AGENT_DATA_EVENT", data)

        case AgentEventConstants.msgError:
            sendEvent("AGENT_ERROR", [
                "code": event.s1 ?? "",
                "message": event.s2 ?? ""
            ])

        case AgentEventConstants.msgToolCall:
            sendEvent("AGENT_TOOL_CALL", [
                "toolName": event.s1 ?? "",
                "params": event.s2 ?? ""
            ])

        case AgentEventConstants.msgToolResult:
            sendEvent("AGENT_TOOL_RESULT", [
                "toolName": event.s1 ?? "",
                "success": event.n1 == 1,
                "result": event.s2 ?? ""
            ])

        case AgentEventConstants.msgActiveSpeakers:
            sendEvent("AGENT_ACTIVE_SPEAKERS", ["speakers": event.s1 ?? ""])

        case AgentEventConstants.msgToolExecuted:
            sendEvent("AGENT_TOOL_EXECUTED", ["data": event.s1 ?? ""])

        case AgentEventConstants.msgRpcMethodCalled:
            sendEvent("AGENT_RPC_METHOD_CALLED", [
                "method": event.s1 ?? "",
                "payload": event.s2 ?? ""
            ])

        default:
            break
        }
    }
}

File: mobile-app/modules/sdk/android/src/main/java/expo/modules/glasssdk/agent/AgentModule.kt
package expo.modules.glasssdk.agent

import android.nfc.Tag
import android.os.Bundle
import android.util.Log
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.glasssdk.tools.ToolManager
import expo.modules.glasssdk.glass.GlassEventConstants
import expo.modules.glasssdk.glass.GlassEventMsg
import expo.modules.glasssdk.network.ApiClient
import expo.modules.glasssdk.SdkConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * Expo module for voice agent control
 * Thin layer that forwards calls to AgentManager and emits events to JS
 */
class AgentModule : Module() {

    companion object {
        private const val TAG = "[Agent:Module]"
        // Duration of inactivity before voice wakeup audio session is auto-disabled
        private const val WAKEUP_AUDIO_TIMEOUT_MS = 10_000L
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private lateinit var agentManager: AgentManager
    private var wakeupInactivityJob: Job? = null
    private var isWakeupSessionActive: Boolean = false

    override fun definition() = ModuleDefinition {
        Name("AgentModule")

        Events(
            "AGENT_STATE_CHANGED",
            "AGENT_PARTICIPANT_EVENT",
            "AGENT_AUDIO_TRACK_EVENT",
            "AGENT_TEXT_EVENT",
            "AGENT_DATA_EVENT",
            "AGENT_ERROR",
            "AGENT_TOOL_CALL",
            "AGENT_TOOL_RESULT",
            "AGENT_ACTIVE_SPEAKERS",
            "AGENT_TOOL_EXECUTED",
            "AGENT_RPC_METHOD_CALLED"
        )

        OnCreate {
            // Initialize AgentManager
            val context = appContext.reactContext!!.applicationContext
            SdkConfig.init(context)
            ApiClient.getInstance(SdkConfig.apiUrl)
            agentManager = AgentManager.getInstance(context)
            agentManager.init()
            ToolManager.getInstance(appContext.reactContext!!);
            // Register for EventBus events
            EventBus.getDefault().register(this@AgentModule)
        }

        OnDestroy {
            // Unregister from EventBus
            EventBus.getDefault().unregister(this@AgentModule)
            wakeupInactivityJob?.cancel()
            wakeupInactivityJob = null
            isWakeupSessionActive = false
            agentManager.destroy()
        }

        OnActivityEntersForeground {
            ToolManager.getInstance()?.setActivity(appContext.currentActivity)
        }

        OnActivityEntersBackground {
            ToolManager.getInstance()?.setActivity(null)
        }

        /**
         * Start the voice agent - connects to LiveKit
         */
        AsyncFunction("start") { audio: Boolean, promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Starting voice agent (audio=$audio)")
                    agentManager.start(audio)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to start agent", e)
                    promise.reject("START_FAILED", e.message, e)
                }
            }
        }

        /**
         * Stop the voice agent
         */
        AsyncFunction("stop") { promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Stopping voice agent")
                    agentManager.stop()
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to stop agent", e)
                    promise.reject("STOP_FAILED", e.message, e)
                }
            }
        }

        /**
         * Send text message to agent
         */
        AsyncFunction("sendText") { text: String, promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Sending text: $text")
                    agentManager.sendText(text)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to send text", e)
                    promise.reject("SEND_TEXT_FAILED", e.message, e)
                }
            }
        }

        /**
         * Send file to agent
         */
        AsyncFunction("sendFile") { path: String, promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Sending file: $path")
                    agentManager.sendFile(path)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to send file", e)
                    promise.reject("SEND_FILE_FAILED", e.message, e)
                }
            }
        }

        /**
         * Send binary data to agent
         */
        AsyncFunction("sendData") { data: ByteArray, reliable: Boolean, promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Sending data: ${data.size} bytes (reliable=$reliable)")
                    agentManager.sendData(data, reliable)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to send data", e)
                    promise.reject("SEND_DATA_FAILED", e.message, e)
                }
            }
        }

        /**
         * Enable/disable microphone
         */
        AsyncFunction("setMicEnabled") { enabled: Boolean, promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Setting mic enabled: $enabled")
                    agentManager.setMicEnabled(enabled)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to set mic enabled", e)
                    promise.reject("SET_MIC_FAILED", e.message, e)
                }
            }
        }

        /**
         * Enable/disable speaker
         */
        AsyncFunction("setSpeakerEnabled") { enabled: Boolean, promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Setting speaker enabled: $enabled")
                    agentManager.setSpeakerEnabled(enabled)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to set speaker enabled", e)
                    promise.reject("SET_SPEAKER_FAILED", e.message, e)
                }
            }
        }

        /**
         * Toggle audio (both mic and agent audio)
         */
        AsyncFunction("toggleAudio") { promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Toggling audio")
                    agentManager.toggleAudio()
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to toggle audio", e)
                    promise.reject("TOGGLE_AUDIO_FAILED", e.message, e)
                }
            }
        }

        /**
         * Set audio enabled/disabled
         */
        AsyncFunction("setAudioEnabled") { enabled: Boolean, promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Setting audio enabled: $enabled")
                    agentManager.setAudioEnabled(enabled)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to set audio enabled", e)
                    promise.reject("SET_AUDIO_FAILED", e.message, e)
                }
            }
        }

        /**
         * Get current connection state
         */
        AsyncFunction("getConnectionState") { ->
            try {
                val state = agentManager.getConnectionState()
                return@AsyncFunction state.name
            } catch (e: Exception) {
                Log.e(TAG, "Failed to get connection state", e)
                throw e
            }
        }

        /**
         * Get microphone enabled state
         */
        AsyncFunction("getMicEnabled") { ->
            try {
                val enabled = agentManager.getMicEnabled()
                return@AsyncFunction enabled
            } catch (e: Exception) {
                Log.e(TAG, "Failed to get mic enabled", e)
                throw e
            }
        }

        /**
         * Get speaker enabled state
         */
        AsyncFunction("getSpeakerEnabled") { ->
            try {
                val enabled = agentManager.getSpeakerEnabled()
                return@AsyncFunction enabled
            } catch (e: Exception) {
                Log.e(TAG, "Failed to get speaker enabled", e)
                throw e
            }
        }

        /**
         * Get room ID
         */
        AsyncFunction("getRoomId") { ->
            try {
                val roomId = agentManager.getRoomId()
                return@AsyncFunction roomId
            } catch (e: Exception) {
                Log.e(TAG, "Failed to get room ID", e)
                throw e
            }
        }

        AsyncFunction("reloadTools") { promise: Promise ->
            scope.launch {
                try {
                    agentManager.reloadTools()
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to reload tools", e)
                    promise.reject("RELOAD_TOOLS_FAILED", e.message, e)
                }
            }
        }

        AsyncFunction("sendFiles") { urls: List<String>, message: String, promise: Promise ->
            scope.launch {
                try {
                    agentManager.sendFiles(urls, message)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to send files", e)
                    promise.reject("SEND_FILES_FAILED", e.message, e)
                }
            }
        }

        AsyncFunction("performRpc") { method: String, payload: String, promise: Promise ->
            scope.launch {
                try {
                    val result = agentManager.performRpc(method, payload)
                    promise.resolve(result)
                } catch (e: Exception) {
                    promise.reject("PERFORM_RPC_FAILED", e.message ?: "Unknown error", e)
                }
            }
        }

        AsyncFunction("registerRpcMethod") { method: String, promise: Promise ->
            scope.launch {
                try {
                    agentManager.registerRpcMethod(method)
                    promise.resolve(null)
                } catch (e: Exception) {
                    promise.reject("REGISTER_RPC_METHOD_FAILED", e.message ?: "Unknown error", e)
                }
            }
        }

        /**
         * Start an app session
         */
        AsyncFunction("startApp") { appId: String, promise: Promise ->
            scope.launch {
                try {
                    Log.d(TAG, "Starting app: $appId")
                    agentManager.startApp(appId)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to start app", e)
                    promise.reject("START_APP_FAILED", e.message, e)
                }
            }
        }

        /**
         * Set authentication cookie for API requests
         */
        AsyncFunction("setCookie") { cookie: String, promise: Promise ->
            try {
                Log.d(TAG, "Setting cookie")
                agentManager.setCookie(cookie)
                promise.resolve(true)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to set cookie", e)
                promise.reject("SET_COOKIE_FAILED", e.message, e)
            }
        }

        AsyncFunction("startPhoneStream") { fps: Int, bitrate: Int, promise: Promise ->
            scope.launch {
                try {
                    agentManager.startPhoneStream(fps, bitrate)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to start phone stream", e)
                    promise.reject("START_PHONE_STREAM_FAILED", e.message, e)
                }
            }
        }

        AsyncFunction("stopPhoneStream") { promise: Promise ->
            scope.launch {
                try {
                    agentManager.stopPhoneStream()
                    promise.resolve(null)
                } catch (e: Exception) {
                    promise.reject("STOP_PHONE_STREAM_FAILED", e.message, e)
                }
            }
        }

        AsyncFunction("flipPhoneCamera") { promise: Promise ->
            try {
                agentManager.flipPhoneCamera()
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject("FLIP_PHONE_CAMERA_FAILED", e.message, e)
            }
        }

        AsyncFunction("isPhoneStreaming") { promise: Promise ->
            promise.resolve(agentManager.isPhoneStreaming())
        }

        /**
         * Set agent name for token dispatch (empty string = use default)
         */
        AsyncFunction("startGlassStream") { ssid: String, pwd: String, fps: Int, bitrate: Int, promise: Promise ->
            scope.launch {
                try {
                    agentManager.startGlassStream(ssid, pwd, fps, bitrate)
                    promise.resolve(null)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to start glass stream", e)
                    promise.reject("START_GLASS_STREAM_FAILED", e.message, e)
                }
            }
        }

        AsyncFunction("stopGlassStream") { promise: Promise ->
            try {
                agentManager.stopGlassStream()
                promise.resolve(null)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to stop glass stream", e)
                promise.reject("STOP_GLASS_STREAM_FAILED", e.message, e)
            }
        }

        AsyncFunction("isGlassStreaming") { promise: Promise ->
            promise.resolve(agentManager.isGlassStreaming())
        }

        AsyncFunction("setAgentName") { name: String, promise: Promise ->
            try {
                Log.d(TAG, "Setting agent name: $name")
                agentManager.setAgentName(if (name.isEmpty()) null else name)
                promise.resolve(null)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to set agent name", e)
                promise.reject("SET_AGENT_NAME_FAILED", e.message, e)
            }
        }
    }

    /**
     * Subscribe to Glass events from EventBus — handles voice wakeup to manage audio session
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onGlassEvent(event: GlassEventMsg) {
        when (event.cmd) {
            GlassEventConstants.MSG_GLASS_VOICE_WAKEUP -> handleVoiceWakeup()
            GlassEventConstants.MSG_GLASS_VOICE_STOP -> handleVoiceWakeupEnd()
            GlassEventConstants.MSG_GLASS_CONNECTED,
            GlassEventConstants.MSG_GLASS_DISCONNECTED -> agentManager.syncGlassAiStatus()
        }
    }

    /**
     * On voice wakeup: enable audio if agent is connected but audio is off,
     * then schedule auto-disable after WAKEUP_AUDIO_TIMEOUT_MS of inactivity.
     */
    private fun handleVoiceWakeup() {
        Log.d(TAG, "Voice wakeup received")
        val state = agentManager.getConnectionState()
        if (state != ConnectionState.CONNECTED) {
            Log.d(TAG, "Agent not connected — ignoring voice wakeup")
            return
        }

        // Echo guard: a phone-initiated voice session pushes speaking_start to the glasses, which can
        // make the device echo aiDialogue=true → a spurious wakeup. Ignore it when voice is already on
        // and we're not in a device-initiated wakeup session.
        // Idempotent guard: ignore the wakeup whenever voice is already on — covers repeated
        // aiDialogue=true frames and the speaking_start echo, so we never re-send "hey" / re-enable.
        if (agentManager.getMicEnabled()) {
            Log.d(TAG, "Voice already active — ignoring wakeup")
            return
        }

        // Send text first for lower latency
        scope.launch {
            try {
                agentManager.sendText("hey")
            } catch (e: Exception) {
                Log.w(TAG, "Could not send wakeup message", e)
            }
        }

        if (!agentManager.getMicEnabled()) {
            Log.i(TAG, "Voice wakeup: enabling audio")
            agentManager.setAudioEnabled(true)
        }

        isWakeupSessionActive = true
        // Inactivity auto-off handled by backend (exit_voice_mode) — app-side timer disabled.
        // resetWakeupTimer()
    }

    /**
     * Glasses' button turned voice mode OFF (device AI-dialogue falling edge) → turn off the
     * phone-side voice and cancel the wakeup inactivity timer.
     */
    private fun handleVoiceWakeupEnd() {
        Log.d(TAG, "Voice wakeup end received")
        wakeupInactivityJob?.cancel()
        wakeupInactivityJob = null
        isWakeupSessionActive = false
        if (agentManager.getMicEnabled()) {
            agentManager.setAudioEnabled(false)
        }
    }

    /**
     * Restart the inactivity countdown without re-enabling audio.
     * Called on initial wakeup and whenever conversation activity is detected.
     */
    private fun resetWakeupTimer() {
        wakeupInactivityJob?.cancel()
        wakeupInactivityJob = scope.launch {
            delay(WAKEUP_AUDIO_TIMEOUT_MS)
            Log.i(TAG, "Voice wakeup session timed out after inactivity — disabling audio")
            agentManager.setAudioEnabled(false)
            isWakeupSessionActive = false
            wakeupInactivityJob = null
        }
    }

    /**
     * Subscribe to Agent events from EventBus.
     * Also resets the wakeup inactivity timer when conversation activity is detected.
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onAgentEvent(event: AgentEventMsg) {
        Log.i(TAG, "onAgentEVentxx ${event}")

        // Inactivity auto-off handled by backend (exit_voice_mode) — app-side activity tracking disabled.
        // if (isWakeupSessionActive) {
        //     when (event.cmd) {
        //         AgentEventConstants.MSG_AGENT_ACTIVE_SPEAKERS -> {
        //             val speakers = event.s1
        //             if (!speakers.isNullOrEmpty() && speakers != "[]") {
        //                 Log.d(TAG, "Wakeup: activity detected (active speakers) — resetting timer")
        //                 resetWakeupTimer()
        //             }
        //         }
        //         AgentEventConstants.MSG_AGENT_TEXT_EVENT -> {
        //             Log.d(TAG, "Wakeup: activity detected (transcription) — resetting timer")
        //             resetWakeupTimer()
        //         }
        //     }
        // }

        emitEvent(event)
    }

    /**
     * Emit event to JavaScript
     */
    private fun emitEvent(event: AgentEventMsg) {
        Log.i(TAG, "eventxxxx ${event}")
        try {
            val eventData = Bundle()

            // Map based on event type
            when (event.cmd) {
                AgentEventConstants.MSG_AGENT_STATE_CHANGED -> {
                    eventData.putString("state", event.s1)
                    sendEvent("AGENT_STATE_CHANGED", eventData)
                }

                AgentEventConstants.MSG_AGENT_PARTICIPANT_EVENT -> {
                    eventData.putString("type", event.s1)
                    eventData.putString("participantId", event.s2)
                    event.s3?.let { eventData.putString("identity", it) }
                    event.s4?.let { eventData.putString("name", it) }
                    eventData.putBoolean("isSpeaking", event.n1 == 1)
                    eventData.putBoolean("isAgent", event.n2 == 1)
                    sendEvent("AGENT_PARTICIPANT_EVENT", eventData)
                }

                AgentEventConstants.MSG_AGENT_AUDIO_TRACK_EVENT -> {
                    eventData.putString("type", event.s1)
                    eventData.putString("participantId", event.s2)
                    eventData.putString("trackId", event.s3 ?: "")
                    eventData.putBoolean("isLocal", event.n1 == 1)
                    eventData.putBoolean("isMuted", event.n2 == 1)
                    sendEvent("AGENT_AUDIO_TRACK_EVENT", eventData)
                }

                AgentEventConstants.MSG_AGENT_TEXT_EVENT -> {
                    eventData.putString("id", event.s1)
                    eventData.putString("text", event.s2)
                    eventData.putBoolean("isFinal", event.n1 == 1)
                    eventData.putBoolean("isUser", event.n2 == 1)
                    eventData.putBoolean("isTranscription", event.n3 == 1)
                    event.s3?.let { eventData.putString("segmentId", it) }
                    sendEvent("AGENT_TEXT_EVENT", eventData)
                }

                AgentEventConstants.MSG_AGENT_DATA_EVENT -> {
                    eventData.putString("participantId", event.s1)
                    eventData.putString("data", event.s2) // Base64 encoded
                    event.s3?.let { eventData.putString("topic", it) }
                    sendEvent("AGENT_DATA_EVENT", eventData)
                }

                AgentEventConstants.MSG_AGENT_ERROR -> {
                    eventData.putString("code", event.s1)
                    eventData.putString("message", event.s2)
                    sendEvent("AGENT_ERROR", eventData)
                }

                AgentEventConstants.MSG_AGENT_TOOL_CALL -> {
                    eventData.putString("toolName", event.s1)
                    eventData.putString("params", event.s2)
                    sendEvent("AGENT_TOOL_CALL", eventData)
                }

                AgentEventConstants.MSG_AGENT_TOOL_RESULT -> {
                    eventData.putString("toolName", event.s1)
                    eventData.putBoolean("success", event.n1 == 1)
                    eventData.putString("result", event.s2)
                    sendEvent("AGENT_TOOL_RESULT", eventData)
                }

                AgentEventConstants.MSG_AGENT_ACTIVE_SPEAKERS -> {
                    eventData.putString("speakers", event.s1)
                    sendEvent("AGENT_ACTIVE_SPEAKERS", eventData)
                }

                AgentEventConstants.MSG_AGENT_TOOL_EXECUTED -> {
                    eventData.putString("data", event.s1)
                    sendEvent("AGENT_TOOL_EXECUTED", eventData)
                }

                AgentEventConstants.MSG_AGENT_RPC_METHOD_CALLED -> {
                    eventData.putString("method", event.s1)
                    eventData.putString("payload", event.s2)
                    sendEvent("AGENT_RPC_METHOD_CALLED", eventData)
                }
            }

            Log.d(TAG, "Emitted Agent event: ${event.cmd}")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to emit Agent event: ${event.cmd}", e)
        }
    }
}
